### ISSUE TITLE
Caffeine will not disable cinnamon-screensaver

### ISSUE THREAD

**Issue by vasek125** [linuxmint/Cinnamon#2458]
Caffeine (https://launchpad.net/caffeine) does not work properly - it should disable screensaver while running some application - e.g. smplayer. It detects smplayer but screensaver will not be disabled.

**Comment by cayjd** [comment]
The problem is still present in Cinnamon 17.1, using the latest Caffeine (2.8.3).

**Comment by ACinnamonUserInUK** [comment]
As the previous poster says, this problem persists (even, in fact, when using the nightly builds of Cinnamon). This irritation will affect many people. Surely someone should be working on it.

There's another report of this problem here: https://github.com/linuxmint/cinnamon-screensaver/issues/17

and here:

https://bugs.launchpad.net/caffeine/+bug/1026411?comments=all

**Comment by bluetiger9** [comment]
Same with Linux Mint Cinnamon 17.3 and Caffeine 2.8.3.
Version 2.6 works as expected.

**Comment by mtwebster** [comment]
Looking here: http://bazaar.launchpad.net/~caffeine-developers/caffeine/main/view/head:/caffeine-screensaver

This file does a number of things to decide which way to suspend the screensaver.

Starting here on line 934: http://bazaar.launchpad.net/~caffeine-developers/caffeine/main/view/head:/caffeine-screensaver#L930

It runs the detectDE function, from which it determines the de is "gnome" based on the fact that something is providing `org.gnome.SessionManager` - our cinnamon-session does.  This is all great and fine (we left it that name precisely because of inhibit functions like this).

The problem is, it gets to line 934, and if it sees that gnome-screensaver-command is installed, it changes its DE to 'gnome_screensaver', which results in it trying to communicate directly with gnome-screensaver over dbus.

If you **remove** gnome-screensaver package, this script works fine in cinnamon, caffeine should inhibit properly.

I'm not sure why this is necessary for them - they can inhibit the screensaver just as easily using org.gnome.SessionManager (which, in turn, talks to org.gnome.ScreenSaver).

It seems to me the correct thing to to is to remove line 934 from that script, and its associated code.  It looks unnecessary (unless gnome has changed their gnome-session stuff drastically that I'm unaware of).  Or they need to add special handling for cinnamon (like checking for cinnamon-screensaver-command).

tldr,

Try removing gnome-screensaver, or remove line 934 (http://bazaar.launchpad.net/~caffeine-developers/caffeine/main/view/head:/caffeine-screensaver#L934)

**Comment by mtwebster** [comment]
btw, the new inhibit applet in cinnamon 2.8 will, in addition to allowing the user to inhibit power management, show if *another* program is already inhibiting.  I mention this only for verification purposes if anyone tries to troubleshoot this further.

**Comment by bluetiger9** [comment]
Actually the `gnome-screensaver` package is not installed, at least in Linux Mint 17.3. The `gnome-screensaver-command` command is a symbolic link to `cinnamon-screensaver-command`:
```
~ $ whereis gnome-screensaver-command
gnome-screensaver-command: /usr/bin/gnome-screensaver-command /usr/bin/X11/gnome-screensaver-command
~ $ readlink /usr/bin/gnome-screensaver-command
/usr/bin/cinnamon-screensaver-command
```

Seems that the Cinnamon screensaver has the DBus interface too, but the service / path / interface names were changed from `org.gnome.ScreenSaver` to `org.cinnamon.ScreenSaver`: https://github.com/linuxmint/cinnamon-screensaver/commit/98c20198b53ec82ea8a602d289539809442ac33a

I think the solution would be to change the line 934 either to search the DBus interface instead of the gnome-screensaver-command (`dbus-send --print-reply --dest=org.freedesktop.DBus /org/freedesktop/DBus org.freedesktop.DBus.GetNameOwner string:org.gnome.ScreenSaver`)
or the explicitly detect the Cinnamon DBus interface and use that with the appropriate names.

As for the `DE=gnome` the some commands (`activate|lock|reset`) are not implemented, the second solution seems better one.

Created an issue on the Caffeine project:
https://bugs.launchpad.net/caffeine/+bug/1535846

**Comment by mtwebster** [comment]
Strange - i have no such symlink on my system.  That would definitely mess this script up.

Using the freedesktop interface is definitely the best way to go I think, but as you say, it doesn't support every command.  I've not actually used or look further into caffeine to see if that would be an actual issue or not.

**Comment by mtwebster** [comment]
In the meantime, I'm looking into why we link to those names - it may not be necessary anymore for us.

**Comment by mtwebster** [comment]
FYI, https://github.com/linuxmint/cinnamon-screensaver/commit/058b9ecac579ebf47f083256f803d89999b73764

cinnamon-screensaver will no longer create gnome-screensaver symbolic links if gnome-screensaver isn't installed.  This should cause the caffeine-screensaver script to work correctly and use the dbus interface (so long as gnome-screensaver is not actually installed, of course).
